### ISSUE TITLE
Flutter web's TextField displays gibberish for Unicode inputs, such as Korean characters (CJK)

### ISSUE THREAD

**Issue by wjkoh** [flutter/flutter#138288]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

To reproduce the bug, create a Flutter project with a TextField, run the app for Web (`flutter run -d chrome`), switch to Korean keyboard, and type Korean characters into the text field. This happens to emojis (Ctrl-Cmd-Space) as well on Mac OS.

### Expected results

 I expected the typed Korean alphabet characters to show correctly, like Latin alphabets do.

### Actual results

However, the typed Korean characters display as gibberish (a box with an x mark) for a few seconds before turning into the correct Korean characters. Check out the attached video below.

### Code sample

<details><summary>Code sample</summary>

```dart
TextField();
```

</details>

### Screenshots or Video

<details>
<summary>Screenshots / Video demonstration</summary>

https://github.com/flutter/flutter/assets/1053445/923a5a31-f70e-4c92-b8c1-e7cd65ed09e0

</details>

### Logs

<details><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
$ flutter doctor -v
[✓] Flutter (Channel stable, 3.13.9, on macOS 13.5.2 22G91 darwin-arm64, locale en-US)
    • Flutter version 3.13.9 on channel stable at /opt/homebrew/Caskroom/flutter/3.13.5/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision d211f42860 (2 weeks ago), 2023-10-25 13:42:25 -0700
    • Engine revision 0545f8705d
    • Dart version 3.1.5
    • DevTools version 2.25.0

[✗] Android toolchain - develop for Android devices
    ✗ Unable to locate Android SDK.
      Install Android Studio from: https://developer.android.com/studio/index.html
      On first launch it will assist you in installing the Android SDK components.
      (or visit https://flutter.dev/docs/get-started/install/macos#android-setup for detailed instructions).
      If the Android SDK has been installed to a custom location, please use
      `flutter config --android-sdk` to update to that location.


[✓] Xcode - develop for iOS and macOS (Xcode 15.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 15A507
    • CocoaPods version 1.12.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[!] Android Studio (not installed)
    • Android Studio not found; download from https://developer.android.com/studio/index.html
      (or visit https://flutter.dev/docs/get-started/install/macos#android-setup for detailed instructions).

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-arm64   • macOS 13.5.2 22G91 darwin-arm64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 119.0.6045.123

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 2 categories.
```

</details>

**Comment by huycozy** [comment]
Thanks for filing the issue. Reproduced this issue on Web platform whilst it doesn't happen on desktop platform (checked on macOS).

<details>
<summary>flutter doctor -v (stable and master)</summary>

```bash
[✓] Flutter (Channel stable, 3.13.9, on macOS 14.1 23B74 darwin-x64, locale en-VN)
    • Flutter version 3.13.9 on channel stable at /Users/huynq/Documents/GitHub/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision d211f42860 (11 days ago), 2023-10-25 13:42:25 -0700
    • Engine revision 0545f8705d
    • Dart version 3.1.5
    • DevTools version 2.25.0

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/huynq/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • ANDROID_HOME = /Users/huynq/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.0.1)
    • Xcode at /Users/huynq/Downloads/Xcode.app/Contents/Developer
    • Build 15A507
    • CocoaPods version 1.13.0

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.3)
    • Android Studio at /Applications/Android Studio Giraffe.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b829.9-10027231)

[✓] Android Studio (version 2022.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • android-studio-dir = /Applications/Android Studio.app/
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)

[✓] VS Code (version 1.84.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.76.0

[✓] Connected device (4 available)
    • RMX2001 (mobile) • EUYTFEUSQSRGDA6D                         • android-arm64  • Android 11 (API 30)
    • iPhone (mobile)  • d9a94afe2b649fef56ba0bfeb052f0f2a7dae95e • ios            • iOS 15.7.9 19H365
    • macOS (desktop)  • macos                                    • darwin-x64     • macOS 14.1 23B74 darwin-x64
    • Chrome (web)     • chrome                                   • web-javascript • Google Chrome 118.0.5993.117

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

```bash
[!] Flutter (Channel master, 3.17.0-5.0.pre.64, on macOS 14.1 23B74 darwin-x64, locale en-VN)
    • Flutter version 3.17.0-5.0.pre.64 on channel master at /Users/huynq/Documents/GitHub/flutter_master
    ! Warning: `flutter` on your path resolves to /Users/huynq/Documents/GitHub/flutter/bin/flutter, which is not inside your current Flutter SDK checkout at /Users/huynq/Documents/GitHub/flutter_master. Consider adding /Users/huynq/Documents/GitHub/flutter_master/bin to the front of your path.
    ! Warning: `dart` on your path resolves to /Users/huynq/Documents/GitHub/flutter/bin/dart, which is not inside your current Flutter SDK checkout at /Users/huynq/Documents/GitHub/flutter_master. Consider adding /Users/huynq/Documents/GitHub/flutter_master/bin to the front of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 2e088178e8 (4 hours ago), 2023-11-12 18:45:26 -0500
    • Engine revision a18ee3c7f5
    • Dart version 3.3.0 (build 3.3.0-120.0.dev)
    • DevTools version 2.30.0-dev.1
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/huynq/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • ANDROID_HOME = /Users/huynq/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 15A507
    • CocoaPods version 1.13.0

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.3)
    • Android Studio at /Applications/Android Studio Giraffe.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b829.9-10027231)

[!] Android Studio (version unknown)
    • Android Studio at /Applications/Android Studio Preview.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    ✗ Unable to determine Android Studio version.
    • Java version OpenJDK Runtime Environment (build 17.0.8+0-17.0.8b1000.22-10799086)

[✓] Android Studio (version 2022.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • android-studio-dir = /Applications/Android Studio.app/
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)

[✓] VS Code (version 1.84.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.76.0

[✓] Connected device (3 available)
    • RMX2001 (mobile) • EUYTFEUSQSRGDA6D • android-arm64  • Android 11 (API 30)
    • macOS (desktop)  • macos            • darwin-x64     • macOS 14.1 23B74 darwin-x64
    • Chrome (web)     • chrome           • web-javascript • Google Chrome 119.0.6045.123

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 2 categories.
```
</details>

**Comment by htoor3** [comment]
@wjkoh Once it correctly displays the korean characters, does it ever go back to the broken characters? And does it happen every time you load the page, or does waiting a bit longer before typing the korean characters fix it? This seems like a font loading issue

**Comment by wjkoh** [comment]
@htoor3 No, it doesn't revert the broken characters. Yes, it occurs every time I reload the webpage. No, it doesn't matter how much time I wait before I begin typing. You can duplicate the behavior by utilizing [this DartPad link](https://api.flutter.dev/flutter/material/TextField-class.html#material.TextField.2) and copying and pasting the text: "가나다라마바사" or "👍". The broken character also appears when typing emojis.

**Comment by wjkoh** [comment]
This issue also occurs when displaying non-Latin alphabet characters, rather than typing them. For instance, when using `Text("가나다라 😆")` instead of `TextField(...)` broken characters can be seen for a few seconds, even in the release build.

**Comment by yjbanov** [comment]
This is a result of lazy-loading fonts. A workaround is to add the necessary font in your app's manifest, although the app will pay an upfront cost to fetch the font before rendering anything.

Long-term we'd like to optimize it, either by implementing platform-driven text fields, or by optimizing our font loading system.

**Comment by yjbanov** [comment]
See also https://github.com/flutter/flutter/issues/120613

**Comment by wjkoh** [comment]
@yjbanov Thanks for the clarification! Do you have a user guide for the workaround you mentioned? This is a very serious and release-blocking issue for any international app, as users will see the app covered with broken characters for a few seconds each time they open the app.

**Comment by hillzy911** [comment]
> @yjbanov Thanks for the clarification! Do you have a user guide for the workaround you mentioned? This is a very serious and release-blocking issue for any international app, as users will see the app covered with broken characters for a few seconds each time they open the app.

How did you work around this @wjkoh ?

**Comment by wjkoh** [comment]
@hillzy911 We were unable to find a solution and have since discontinued the use of Flutter for our consumer application.

**Comment by Mr-Pepe** [comment]
I am working on an online Chinese-English dictionary implemented in Flutter (https://app.chill-chinese.com) and this issue is something I struggled with a lot. 

Without any cached content, a user had to download 6MB of fonts, and 3.5MB of dictionary data, before being able to use the app. Having the downloads run in parallel is pretty bad on slower connections because neither the display nor the search work until everything is loaded.

I ended up creating a font subset and added it to my pubspec.yaml. It contains only the characters I know I have to load for the initial screen to show. However, the font gets loaded before the first frame and thus blocks the app startup, because https://github.com/flutter/flutter/issues/122282 is not implemented yet. But at least it's only 280KB in size.

A font subset that covers all characters in my app would be smaller than 6MB, but there is currently no way to load it on demand, so it would still block the app startup for too long.

Therefore, eventually, the full 6MB font needs to get loaded to display all characters throughout the app. My workaround to proactively trigger the download is to render a character that is not part of my font subset in a stack behind the actual UI.

@hillzy911 Maybe that's something that works for you?

Conclusion: Sliced fonts for CJK characters would be great, so the user doesn't have to wait for a 6MB download. Having more control over how custom fonts are loaded (#122282) would also be great, so developers could craft their own font subsets and load them proactively whenever they think is a good time during app startup.
